<a href="https://colab.research.google.com/github/udayameister/PythonTrainingProgramme/blob/main/Week_2_Python_String_Tutorial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python String Tutorial - Python String Function & Operations

## 1. Python String Tutorial

#### We will learn how to declare and slice a string in python and also look at the python string functions and python string operation.

#### At last, we cover escape sequences in Python. As we saw earlier, you dont need to mention the data type when declaring a string

## 2. What is Python String?

#### A Python string is a sequence of characters. There is a built-in class 'str' for handling Python string. You can prove this with type() function.|

In [ ]:
type('Dogs are love')

str

## 3. How to Declare Python String?

#### You can declare a python string using either single quotes or double quotes.

In [ ]:
a='Dogs are love'

In [ ]:
print(a)

Dogs are love


In [ ]:
a = "Dogs are love"

In [ ]:
print(a)

Dogs are love


#### However, you cannot use a single quote to begin a string and a double quote to end it, and vice-versa

## 5. Spanning a string across lines

#### When you want to span a python string across multiple lines, you can use triple quotes

In [ ]:
<pre data-enlighter-language="php" class = "EnlighterJSRAW">>>>
a="""Hello
    welcome"""

SyntaxError: invalid syntax (<ipython-input-6-e3ee1de82af9>, line 1)

In [ ]:
<pre data-enlighter-language="php" class="EnlighterJSRAW">>>>
a="""Hello
    welcome"""
>>> print(a)</pre>

SyntaxError: invalid syntax (<ipython-input-8-1a6fc527a8b6>, line 1)

In [ ]:
a="""Hello
    welcome"""

In [ ]:
print(a)

Hello
    welcome


## 6. How to access python string

#### A string is immutable; it cant be changed.

In [ ]:
a="Dogs are love"

In [ ]:
a[0] ="H"

TypeError: 'str' object does not support item assignment

### a. Displaying a single character

#### To display a single character from a string, put its index in square brackets. Indexing begins at 0

In [ ]:
a[0]

'D'

#### b. Slicing a string

#### Sometimes, you may want to display only a part of a string. For this, use the slicing operator[].

In [ ]:
a[3:8]

's are'

#### Here, it printed characters 3 to 7 with indexing begining at 0.

In [ ]:
a[:8]

'Dogs are'

#### This prints characters from the begining to character 7

In [ ]:
a[8:]

' love'

#### This prints characters from character 8 to the end of the string.

In [ ]:
a[:]

'Dogs are love'

#### This prints the whole string.

In [ ]:
a[:-2]
# this prints characters from the beginning to two characters less than the end of the string

'Dogs are lo'

#### This prints characters from three characters from the strings end to two characters from it.

In [ ]:
a[-3:-2]

'o'

## 7. Python String concatenation

#### Concatenation is the operation of joining stuff together. Python strings can join using the concatenation operator +

In [ ]:
a = 'Do you see this, '
b = '$$?'
a+b

'Do you see this, $$?'

In [ ]:
a = '10'
print(2*a)

1010


#### Multiplying 'a' by 2 returned 1010, and not 20, because '10' is a string, not a number. You can either use commas, or use string formatters for the same.

## 8. Python String Formatters

In [ ]:
city = 'Ahmedabad'

In [ ]:
print ("age",21, "city" , city)

age 21 city Ahmedabad


### I. f-strings

#### The letter 'f' precedes the string, and the variables are mentioned in curly braces in their places.

In [ ]:
name = 'Manish'

In [ ]:
print(f"it isnt {name}'s birthday")

it isnt Manish's birthday


#### Notice that because we wanted to use two single quotes in the string, we delimited the entire string with double quotes instead.

### ii. format() method

#### We can use the format() method to do the same. It succeeds the string, and has the variables as arguments separated by commas. In the string, use curly braces to posit the variables. Inside the curly braces, you can either put 0,1, ... or the variables. When doing the latter, you must assign values to them in the format method.

In [ ]:
print("I have {0}".format(a))

I have 10


In [ ]:
print("I love {a}".format(a='cats'))

I love cats


#### The variables dont have to defined before the print statement

In [ ]:
print("I love {b}".format(b='ferrets'))

I love ferrets


In [ ]:
b

'$$?'

### c. % operator

#### The % operator goes where the variables go in a string. %s is for string. What follows the string is the operator and variables in parentheses/in a tuple.

In [ ]:
a="dog"

In [ ]:
b="ferrets"

In [ ]:
print("I love %s and %s" %(a,b))

I love dog and ferrets


#### Other options include

#### %d - for integers

#### %f - for floating-point numbers

## 9. Escape Sequences in Python

#### In a python string, you may want to put a tab, a linefeed, or other such things. Escape sequences allow us to do this. An escape sequeunce is a backslash followed by a character, depending on what you want to do. Python supports the following sequences.

#### \n - linefeed

#### \t - tab

####  \\\ --backslash

#### Since a blackslash may be a part of an escaped by a backslash. this lets you use single quotes freely in a string

#### \\' single quote can be escaped by a backslash. This lets you use single quotes freely in a string

#### \\" like the single quote, the double quote can be escaped too.

## 10. Python string functions

#### Python provides us with a number of functions that we can apply on strings or to create strings

#### a. len() --> the len() function returns the length of a string.

In [ ]:
a='book'

In [ ]:
len(a)

4

In [ ]:
len(a[2:])

2

#### b. str() --> this function converts any data type into a string

In [ ]:
str(2+3)

'5'

In [ ]:
str(2+3j)

'(2+3j)'

In [ ]:
str(['red','green','blue'])

"['red', 'green', 'blue']"

In [ ]:
len(str(['red','green','blue']))

24

### c. lower() and upper()

In [ ]:
a = 'Book'

In [ ]:
a.lower()

'book'

In [ ]:
a.upper()

'BOOK'

### d. strip() --> removes whitespaces from the begining and end of the string

In [ ]:
b = '  Book '

In [ ]:
a.strip()

'Book'

### e. isdigit() --> return true if all characters in a string are digits.

In [ ]:
c = '777'

In [ ]:
c.isdigit()

True

In [ ]:
d = '77a'

In [ ]:
d.isdigit()

False

### f. isalpha()

#### Returns True if all characters in a string are characters from an alphabet.

In [ ]:
e= 'abc'

In [ ]:
e.isalpha()

True

### g. isspace()

#### Returns True if all characters in a string are spaces.

In [ ]:
a = ' '

In [ ]:
type(a)

str

In [ ]:
a.isspace()

True

In [ ]:
c

'777'

In [ ]:
d

'77a'

In [ ]:
e

'abc'

In [ ]:
f = 'aaa  '

In [ ]:
f.isspace()

False

### h. startswith()

#### It takes a string as an argument, and returns True is the string it is applied on begins with the string in the arguments

In [ ]:
e.startswith('a')

True

### i. endswith()

#### it takes a string as an argument, and returns True if the string it is applied on ends with the string in the arguments

In [ ]:
a='therefore'

In [ ]:
a.endswith('fore')

True

### j. find()

### it takes an argument and searches for it in the string on which it is applied. it then returns the index of the substring.

In [ ]:
'homeowner'.find('meow')

2

In [ ]:
'homeowner'.find('wow')

-1

#### if the string doenst exist in the main string, then the index it returns is -1

### k . replace()

#### It takes two arguments. the first is the substring to be replaced. The second is the substring to replace with

In [ ]:
'banana'.replace('na','ha')

'bahaha'

### I split()

#### It takes one argument. The string is then split around every occurrence of the argument in the string

In [ ]:
'No. Okay. Why?'.split('.')

['No', ' Okay', ' Why?']

In [ ]:
values = 'No. Okay. Why?'

In [ ]:
type(values)

str

In [ ]:
new_values = values.split('.')

In [ ]:
new_values

['No', ' Okay', ' Why?']

In [ ]:
type(new_values)

list

### m. Join()

#### It takes a list as an argument and joins the elements in the list using the string it is applied on

In [ ]:
"*".join(['red','green','blue'])

'red*green*blue'

In [ ]:
"*".join('red')

'r*e*d'

In [ ]:
"*".join("red")

'r*e*d'

## 11 . Python String Operations

### a.  Comparison

In [ ]:
'hey'<'hi'

True

#### 'hey' is lesser than 'hi' lexicographically (because I comes after e in the dictionary)

In [ ]:
a='check'

In [ ]:
a=='check'

True

In [ ]:
'yes' !='no'

True

### b. Arithmetic

#### Some arithmetic operations can be applied on strings

In [ ]:
'ba'+'na'*2

'banana'

In [ ]:
'less' not in 'helpless'

False

In [ ]:
'less' in 'helpless'

True

### d. Identify

#### Python's identify operators 'is' and 'is not' can be used on strings

In [ ]:
'hey' is 'hi'

False

In [ ]:
'him' is 'hi'

False

In [ ]:
'Yo' is not 'yo'

True

In [ ]:
'Manish' is 'Manish'

True

### e. Logical

#### And, or, and not operators can be applied too.

#### and --> if the value on the left is true it returns the value on the right. Otherwise, the value on the left is false, it return False

In [ ]:
'' and '1'

''

In [ ]:
'1' and ''

''

In [ ]:
'' and ''

''

In [ ]:
'1' and '2'

'2'

In [ ]:
'1' and '1'

'1'

In [ ]:
'2' and '1'

'1'

In [ ]:
'q' and '1'

'1'

In [ ]:
a and b

'  Book '

In [ ]:
a

'check'

In [ ]:
b

'  Book '

#### 2. or - if the value on the left is True, it return True . Otherwise, the value on the right is returned.

#### 3. not - As we said earlier , an empty string has a Boolean value of False

In [ ]:
not('1')

False

In [ ]:
not('')

True